### PR TITLE
[MapEditor] Victory/loss conditions in map editor

### PR DESCRIPTION
**Pull request by Nordsoft91**
Should be merged after https://github.com/vcmi/vcmi/pull/1277

### CODE REVIEW

**Comment by IvanSavenko** [comment]
Not sure why, but diff shows changes from copy-paste branch. Even though it was already merged into develop?

**Comment by Nordsoft91** [comment]
Fixed

File: mapeditor/mapsettings.ui [hunk 1]
```diff
@@ -53,6 +60,174 @@
...
+       <item>
+        <widget class="QGroupBox" name="groupBox">
+         <property name="title">
+          <string>Difficulty</string>
+         </property>
+         <layout class="QHBoxLayout" name="horizontalLayout_2">
+          <item>
+           <widget class="QRadioButton" name="diffRadio1">
+            <property name="text">
+             <string>1</string>
```
Comment (IvanSavenko) [inline mapeditor/mapsettings.ui:72]: Mark them as non-translatable in Designer

File: mapeditor/mapsettings.cpp [hunk 2]
```diff
@@ -64,23 +129,304 @@ MapSettings::MapSettings(MapController & ctrl, QWidget *parent) :
...
+			break;
+	};
+	
+	//victory & loss messages
+	ui->victoryMessageEdit->setText(QString::fromStdString(controller.map()->victoryMessage));
+	ui->defeatMessageEdit->setText(QString::fromStdString(controller.map()->defeatMessage));
+	
+	//victory & loss conditions
+	const std::array<std::string, 8> conditionStringsWin = {
+		"No special victory",
```
Comment (IvanSavenko) [inline mapeditor/mapsettings.cpp:163]: Mark them as translateable - tr or QT_TR_NOOP

File: mapeditor/mapsettings.cpp [hunk 3]
```diff
@@ -108,5 +454,423 @@ void MapSettings::on_pushButton_clicked()
...
+				cond.value = expiredDate(loseValueWidget->text());
+				specialDefeat.onFulfill = VLC->generaltexth->allTexts[254];
+				specialDefeat.trigger = EventExpression(cond);
+				break;
+			}
+				
+			case 3: {
+				EventCondition cond(EventCondition::DAYS_WITHOUT_TOWN);
+				assert(loseValueWidget);
+				cond.value = loseValueWidget->text().toInt();
```
Comment (IvanSavenko) [inline mapeditor/mapsettings.cpp:679]: I assume that vcmi supports custom values for "DAYS_WITHOUT_TOWN"?

Comment (Nordsoft91) [reply mapeditor/mapsettings.cpp:679]: Yes, this one is customizable, so you can try to make a map with heroes only without towns

File: mapeditor/mapsettings.cpp [hunk 4]
```diff
@@ -108,5 +454,423 @@ void MapSettings::on_pushButton_clicked()
...
+			for(int i = 0; i < controller.map()->allowedArtifact.size(); ++i)
+				victoryTypeWidget->addItem(QString::fromStdString(VLC->arth->objects[i]->getNameTranslated()), QVariant::fromValue(i));
+			break;
+		}
+			
+		case 1: { //EventCondition::HAVE_CREATURES
+			victoryTypeWidget = new QComboBox;
+			ui->victoryParamsLayout->addWidget(victoryTypeWidget);
+			for(int i = 0; i < VLC->creh->objects.size(); ++i)
+				victoryTypeWidget->addItem(QString::fromStdString(VLC->creh->objects[i]->getNameTranslated()), QVariant::fromValue(i));
```
Comment (IvanSavenko) [inline mapeditor/mapsettings.cpp:739]: Creatures have singular & plural name, so `getNameTranslated` is explicitly removed for them
```suggestion
				victoryTypeWidget->addItem(QString::fromStdString(VLC->creh->objects[i]->getNameSingularTranslated()), QVariant::fromValue(i));
```

Comment (IvanSavenko) [reply mapeditor/mapsettings.cpp:739]: Or another option, depending on context - use plural name:
```suggestion
				victoryTypeWidget->addItem(QString::fromStdString(VLC->creh->objects[i]->getNamePluralTranslated()), QVariant::fromValue(i));
```

**Comment by kambala-decapitator** [comment]
@Nordsoft91 please make normal merges in the future, not "merge with rebase"

**Pull request merged**